File: bonsai_skills/src/main/java/de/unibi/citec/clf/bonsai/skills/CompareSlot.java
package de.unibi.citec.clf.bonsai.skills;

import de.unibi.citec.clf.bonsai.core.exception.CommunicationException;
import de.unibi.citec.clf.bonsai.core.object.MemorySlotReader;
import de.unibi.citec.clf.bonsai.core.object.MemorySlotWriter;
import de.unibi.citec.clf.bonsai.engine.model.AbstractSkill;
import de.unibi.citec.clf.bonsai.engine.model.ExitStatus;
import de.unibi.citec.clf.bonsai.engine.model.ExitToken;
import de.unibi.citec.clf.bonsai.engine.model.config.ISkillConfigurator;

/**
 * Read from ReadSlot and compare to compareString.
 *
 * <pre>
 *
 * Slots:
 *  ReadSlot: [String] [Read]
 *      -> Memory slot the content will be read from
 *
 * ExitTokens:
 *  success.match:      Entrys are the same
 *  success.mismatch:   Entrys are not the same
 *  error:              ReadSlot was empty
 *
 * Sensors:
 *
 * Actuators:
 *
 * </pre>
 *
 * @author pvonneumanncosel
 */
public class CompareSlot extends AbstractSkill {

    private ExitToken tokenMisMatch;
    private ExitToken tokenMatch;
    private ExitToken tokenError;

    private static final String KEY_COMPARE_STRING = "#_COMPARE_STRING";

    MemorySlotReader<String> readSlot;

    private String slotContent;
    private String compareString;

    @Override
    public void configure(ISkillConfigurator configurator) {
        tokenMatch = configurator.requestExitToken(ExitStatus.SUCCESS().withProcessingStatus("match"));
        tokenMisMatch = configurator.requestExitToken(ExitStatus.SUCCESS().withProcessingStatus("misMatch"));
        tokenError = configurator.requestExitToken(ExitStatus.ERROR());

        readSlot = configurator.getReadSlot("ReadSlot", String.class);
        compareString = configurator.requestValue(KEY_COMPARE_STRING);
    }

    @Override
    public boolean init() {
        try {
            slotContent = readSlot.recall();

            if (slotContent == null) {
                logger.warn("your ReadSlot was empty");
            }

        } catch (CommunicationException ex) {
            logger.fatal("Unable to read from memory: ", ex);
            return false;
        }
        return true;
    }

    @Override
    public ExitToken execute() {
        if (slotContent.equals(compareString)) {
            return tokenMatch;
        }
        return tokenMisMatch;
    }

    @Override
    public ExitToken end(ExitToken curToken) {
        if (curToken.getExitStatus().isSuccess()) {
            if(slotContent == null){
                return tokenError;
            }
        }
        return curToken;
    }
}

File: bonsai_skills/src/main/java/de/unibi/citec/clf/bonsai/skills/slots/CompareSlot.kt
package de.unibi.citec.clf.bonsai.skills.slots

import de.unibi.citec.clf.bonsai.core.exception.CommunicationException
import de.unibi.citec.clf.bonsai.core.`object`.MemorySlotReader
import de.unibi.citec.clf.bonsai.engine.model.AbstractSkill
import de.unibi.citec.clf.bonsai.engine.model.ExitStatus
import de.unibi.citec.clf.bonsai.engine.model.ExitToken
import de.unibi.citec.clf.bonsai.engine.model.config.ISkillConfigurator

/**
 * Read from ReadSlot and compare to given regex. String read from slot is getting formated to lower case before
 * comparing.
 *
 * <pre>
 *
 *
 * Options:
 * #_COMPARE_STRING:         [String]
 *  -> Regex used for comparing
 *
 * Slots:
 * ReadSlot: [String] [Read]
 *  -> Memory slot the content will be read from
 *
 *
 *
 * ExitTokens:
 * success.match:      Regex matched
 * success.mismatch:   Regex didn't match
 * error:              ReadSlot was empty
 *
 * Sensors:
 *
 * Actuators:
 *
</pre> *
 *
 * @author nschmitz
 */
class CompareSlot : AbstractSkill() {
    private lateinit var tokenMisMatch: ExitToken
    private lateinit var tokenMatch: ExitToken

    private var readSlot: MemorySlotReader<String?>? = null

    private var slotContent: String? = null
    private var comparePattern: String? = null

    override fun configure(configurator: ISkillConfigurator) {
        tokenMatch = configurator.requestExitToken(ExitStatus.SUCCESS().withProcessingStatus("match"))
        tokenMisMatch = configurator.requestExitToken(ExitStatus.SUCCESS().withProcessingStatus("misMatch"))


        readSlot = configurator.getReadSlot("ReadSlot", String::class.java)
        comparePattern = configurator.requestValue(KEY_COMPARE_STRING)
    }

    override fun init(): Boolean {
        try {
            slotContent = readSlot?.recall<String?>()

            if (slotContent == null) {
                logger.warn("your ReadSlot was empty")
                return false
            }
        } catch (ex: CommunicationException) {
            logger.fatal("Unable to read from memory: ", ex)
            return false
        }
        if (comparePattern == "'" || comparePattern!!.isEmpty() || comparePattern!!.isBlank()) {
            logger.warn("Your Regex Pattern is empty. Defaulting to match.")
        }
        return true
    }

    override fun execute(): ExitToken {
        val sanitizedSlotContent = slotContent?.lowercase() ?: "unknown"
        logger.debug("Trying to match regex '$comparePattern' with slot-content '$sanitizedSlotContent'")
        if (comparePattern?.toRegex()?.matches(sanitizedSlotContent) ?: true) {
            return tokenMatch
        }
        return tokenMisMatch
    }

    override fun end(curToken: ExitToken?): ExitToken? {
        return curToken
    }

    companion object {
        // Name of option could be better, its terrible right now
        private const val KEY_COMPARE_STRING = "#_COMPARE_STRING"
    }
}
